### PR TITLE
Fix #262, where a default value of drop affects *de*serialization.

### LINKED ISSUE #262: Contrary to the documentation, 'default' affects
Colander has a [page](http://docs.pylonsproject.org/projects/colander/en/latest/null_and_drop.html) devoted to describing Colander's use of Null and Drop values. The product's current behaviour doesn't match that description. Specifically, `default` should only affect serialization—but it currently affects deserialization too, _if_ `default` is set to `colander.drop`.

Consider this code:

``` python
import colander

class Person(colander.MappingSchema):
    name = colander.SchemaNode(
        colander.String()
    )
    age = colander.SchemaNode(
        colander.Int(),
        validator=colander.Range(0, 200)
    )
    hair_color = colander.SchemaNode(
        colander.String(),
        default=u'default brown',
        missing=u'missing brown'
    )

schema = Person()

print('serialized = {}'.format(schema.serialize({
    'name': 'Fred',
    'age': 20
})))

print('deserialized = {}'.format(schema.deserialize({
    'name': 'Fred',
    'age': 20
})))
```

It produces:

``` python
serialized = {'age': '20', 'name': u'Fred', 'hair_color': u'default brown'}
deserialized = {'age': 20, 'name': u'Fred', 'hair_color': u'missing brown'}
```

That's exactly what we expect: `'hair_color'` wasn't passed in, so during serialization we got the default value and during deserialization we got the missing value.

Now let's change `default` to `colander.drop`:

``` python
import colander

class Person(colander.MappingSchema):
    name = colander.SchemaNode(
        colander.String()
    )
    age = colander.SchemaNode(
        colander.Int(),
        validator=colander.Range(0, 200)
    )
    hair_color = colander.SchemaNode(
        colander.String(),
        default=colander.drop,
        missing=u'missing brown'
    )

schema = Person()

print('serialized = {}'.format(schema.serialize({
    'name': 'Fred',
    'age': 20
})))

print('deserialized = {}'.format(schema.deserialize({
    'name': 'Fred',
    'age': 20
})))
```

It produces: 

``` python
serialized = {'age': '20', 'name': u'Fred'}
deserialized = {'age': 20, 'name': u'Fred'}
```

`hair_color` was dropped during serialization—but it was _also_ dropped during deserialization. We should have gotten the `missing` value instead.

The problem seems to be that `Mapping.serialize()` and `Mapping.deserialize()` both call `Mapping._impl()`, and `Mapping._impl()` consults `default`. There's a similar problem in `Sequence`: `Sequence.serialize()` and `Sequence.deserialize()` both call `Sequence._impl()`, and `Sequence._impl()` consults `default`. Deserialization should never consult `default`.

At least where mappings are concerned, this is a very old bug. It was introduced in [this commit](https://github.com/Pylons/colander/commit/d15b256ce4a123fc932f720f5a3303c1a700ab91) in 2014. That commit was an attempt to fix [this bug](https://github.com/Pylons/colander/issues/139). That bug actually describes what I'm trying to do: have a  `default` value of `colander.drop` and a `missing` value of `None`. (The `Sequence` bug was introduced in [this commit](https://github.com/Pylons/colander/commit/b5205e03b1701216dea7dd08284694dd34888266) which appears to have copied the `Mapping` code.)

My current workaround is to use `deferred()` and `bind()` to instantiate one variation of the schema for serialization and another one for deserialization.

### PR DESCRIPTION
**Pull request by michaellenaghan**

### CODE REVIEW

**Comment by michaellenaghan** [comment]
Since issue references in titles don't turn into links, here's one that will: #262.

File: colander/tests/test_colander.py [hunk 1]
```diff
@@ -3405,30 +3487,73 @@ def test_imperative_with_implicit_schema_type(self):
...
+            a = colander.SchemaNode(
+                colander.String(),
+                default='abc',
+                missing=colander.drop
+                )
+        node = MySchema()
+        # NOTE: This test fails--but what *should* it do
+        # when a Sequence has a subnode with a default?
+        # result = node.serialize([])
+        # self.assertEqual(result, ['abc'])
```
Comment (michaellenaghan) [inline colander/tests/test_colander.py]: Yes, but look at the `MappingSchema` test at lines 3409-10:

     result = node.deserialize({})
     self.assertEqual(result, {'a':'abc'})

I'm not keen on the asymmetry.

**Comment by michaellenaghan** [comment]
Anyone minding the store?

File: colander/tests/test_colander.py (hunk 1 above)
Comment (mmerickel) [inline colander/tests/test_colander.py]: I agree with @tseaver that this should just return an empty list unless there is actually an item in the sequence, such as with `colander.null`.

**Pull request merged**